src/smtencoding/FStar.SMTEncoding.Encode.fst

```diff
@@ -1267,22 +1268,48 @@ and encode_sigelt' (env:env_t) (se:sigelt) : (decls_t * env_t) =
                  | _ -> false
              in
              let u_leq_u_k u =
-                universe_leq (N.normalize_universe env_tps u) u_k
+                let u = N.normalize_universe env_tps u in
+                universe_leq u u_k 
              in
              let tp_ok (tp:S.binder) (u_tp:universe) =
                 let t_tp = tp.binder_bv.sort in
                 if u_leq_u_k u_tp
                 then true
-                else let formals, _ = U.arrow_formals t_tp in
-                     let _, _, _, u_formals = TcTerm.tc_binders env_tps formals in
-                     //List.iter (fun u -> BU.print1 "Universe of formal: %s\n" (Print.univ_to_string u)) u_formals;
-                     BU.for_all (fun u_formal -> u_leq_u_k u_formal) u_formals
+                else (
+                   let t_tp = 
+                      N.normalize
+                          [Unrefine; Unascribe; Unmeta;
+                          Primops; HNF; UnfoldUntil delta_constant; Beta]
+                          env_tps t_tp
+                   in
+                   let formals, t = U.arrow_formals t_tp in
+                   let _, _, _, u_formals = TcTerm.tc_binders env_tps formals in
+                   let inj = BU.for_all (fun u_formal -> u_leq_u_k u_formal) u_formals in                  
+                   if inj
+                   then (
+                     match (SS.compress t).n with
+                     | Tm_type u -> 
+                      (* retain injectivity for parameters that are type functions
+                         from small universes (i.e., all formals are smaller than the constructed type)
+                         to a universe <= the universe of the constructed type.
+                         See BugBoxInjectivity.fst *)
+                       u_leq_u_k u
+                     | Tm_name _ -> (* this is a value of another type parameter in scope *)
```

Comment:
This name is another (opened) type parameter? If that type parameter itself is a type in a higher universe, that's ok? As in, returning true here is ok in that case?

Reply:
It looks like that's indeed problematic:
```fstar
type itest3 (t: Type u#3) (a: t) : Type u#1 = | Mk3 : itest3 t a
let test3 : Type u#2 -> Type u#1 = itest3 (Type u#2)
let test3_inhabited f : test3 f = Mk3
let test3_inj f0 f1 : Lemma (requires test3 f0 == test3 f1) (ensures f0 == f1) =
  let Mk3 #_ = test3_inhabited f0 in ()
```